### ISSUE TITLE
Selecting a range of ports to try until an open port is found

### ISSUE THREAD

**Issue by BramVanroy** [microsoft/DeepSpeed#2752]
**Is your feature request related to a problem? Please describe.**

It is related to #191. Specifically [my comment](https://github.com/microsoft/DeepSpeed/issues/191#issuecomment-1404961743) there. I have opened a new issue [as requested](https://github.com/microsoft/DeepSpeed/issues/191#issuecomment-1405002089) to discuss the potential of a new feature that might help to prevent the original issue from happening.

The issue occurs that the host-address is already in use by another user or another process. For instance a node with 4 GPUs but two separate jobs with deepspeed that both use 2 GPUs. So their hostfile is identical. This triggers a "port already in use". 

**Describe the solution you'd like**
I thought that it might be useful to add the option to specify a range of ports (e.g. `16000:25000`) as an environment variable or as a CLI option. DeepSpeed could then retry connecting through that range of ports until a suitable port is found, rather than just exiting after the default port is already taken.

@tjruwase mentions that this might be slow. I do not think that that is a _huge_ issue because the search only has to happen once (at the start of the process) until an open port is found and it does not happen in the current, default, scenario. So the user should be aware that when they specify a range of ports, that time until connection may take longer. Through `socket` in Python, I think it should be possible to write this for many different platforms without relying on e.g. `netstat`, but I am not 100% sure how feasible it is in the larger scope of DeepSpeed.

**Comment by tjruwase** [comment]
@BramVanroy, thanks for creating this new issue and providing your thoughts. Just a clarification, my concern regarding retries is not slowness since you correctly pointed out this only occurs once at the beginning of training. My main concern is how the retry logic will work correctly in terms of recovering from `busy port` exception/error. I am not familiar with the available python support for such features. 

We will discuss internally, but a PR would also be appreciated. Thanks!

**Comment by jeffra** [comment]
@BramVanroy, this is an interesting idea. We've actually tried this before for running some unit tests in parallel on a single node and it worked okay from what I recall. For single node deployments this seems probably fine, especially if the user is opting into it as opposed to a default. I do have concerns about multi-node deployments, it would be entirely possible that two nodes have different port availability. This could mean that worker-0 sees port 16000 is already in use so uses 16001, but then worker-1 has port 16000 available so it attempts to use it. This would cause torch.distributed to fail/hang :(
